### PR TITLE
rust-analyzer: new Portfile submission

### PR DESCRIPTION
**Pull request by fabius**
###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 11.1
Xcode 12.3

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: devel/rust-analyzer/Portfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+master_sites        https://github.com/rust-analyzer/rust-analyzer/archive/
+distname            ${version}
+checksums           rmd160  c0db3ee3e3f6db7a03f8b53fa45b180b760f6504 \
+                    sha256  009ab946848719bb05bcaddab1cf5a75a45c5a92ae6dc31763adb42c3229f7db \
+                    size    1435696
+supported_archs     x86_64 arm64
+
+build {
+    system -W ${worksrcpath} "rm lib/README.md"
+    system -W ${worksrcpath} "cargo build --release"
```
Comment (herbygillot) [inline devel/rust-analyzer/Portfile:30]: The `cargo` portgroup (`PortGroup cargo 1.0`) normally does the `cargo build` step for you, which means after specifying it, you don't have to specify an explicit `build` section.  Did the normal structure not work for a particular reason?

Comment (fabius) [reply devel/rust-analyzer/Portfile:30]: Thanks for the review! First of all: Is there any Documentation on those Portgroups other than the _resources directory? 

The normal method unfortunately did not work as the currently packaged Rust version, for whatever reason, assumes that everything inside the lib/ directory is a directory itself. So cargo looks for lib/README.md/Cargo.toml which obviously does not exist.

Comment (fabius) [reply devel/rust-analyzer/Portfile:30]: fyi this is not an issue on the latest stable rust version

Comment (herbygillot) [reply devel/rust-analyzer/Portfile:30]: Hey @fabius, I'll get back to you - I'd like to take a look at the `lib/REAME.md` issue you called out

Comment (fabius) [reply devel/rust-analyzer/Portfile:30]: > Hey @fabius, I'll get back to you - I'd like to take a look at the `lib/REAME.md` issue you called out

@herbygillot Any news on this one? 

_I got a working Portfile for Rust 1.49/aarch64 if that helps you testing_

**Comment by herbygillot** [comment]
So here is a commit with a Portfile for `rust-analyzer` that's written as per best practices for Rust ports in Macports:

https://github.com/herbygillot/macports-ports/commit/a428e73238db25e2b096b75e9bfd74f38bec1f57

Important points of note:
* **no v prefix in `github.setup`** - the `v` at the end of `github.setup` is the branch or tag prefix.  You would use it if the branch or tag name in the `rust-analyzer` branch looked like "**v**2021-01-18", but since it's plain "2021-01-18", you don't specify any prefix.  This is important because this is what the `github` portgroup uses in order to construct the URLs it needs to fetch the source tarballs.

* **no `master_sites` or `distname`** - MacPorts and the many portgroups know how to fetch the source tarballs for a piece of software once you've specified the name, version and optional version prefix/suffix as per the way they expect.  In this Portfile, you've done this with `github.setup`.  However since the Github releases URL for `rust-analyzer` are "archive" links, we do have to specify `github.tarball_from archive`.

* **weird lib/README.md directory** - you're definitely right that something strange is going on with the "lib/README.md" path - we delete it right after files have been extracted

* **no manual invocation of `cargo build`** - we are using the `cargo` portgroup. It knows how to build Cargo projects.  As long as we provide all the deps in the `cargo.crates` section, it will take care of the rest, so we do not need our own `build` section invoking `cargo` ourselves.

* **`cargo.crates` section** - MacPorts wants to be aware of all the dependency crates for a Cargo project.  This is so that all of the crates can be mirrored to MacPorts mirrors, enabling MacPorts to build the port even in the (unlikely) scenario that the upstream Rust repositories are down. It also allows MacPorts to check the integrity of all the dependencies and ensure that all downloads happen in the "fetch" phase of building a Port.  In this way, when it's time to actually build the source, no downloading needs to happen. This is the opposite of what happens when you directly invoke `cargo build` in a `build` block - `cargo` downloads things on its own outside of MacPorts' awareness and does the build. Generating this section can be admittedly annoying.  I use my own utility to do this, [cargo2ports](https://github.com/herbygillot/cargo2ports). Others have found it useful as well.

You can see examples of other Cargo Portfiles here:
* https://github.com/macports/macports-ports/blob/master/tex/tectonic/Portfile
* https://github.com/macports/macports-ports/blob/master/sysutils/diskus/Portfile
* https://github.com/macports/macports-ports/blob/master/editors/ox/Portfile

**Comment by fabius** [comment]
Cheers man! I rebased the branch on your fixup. 
Edit: `supported_archs: arm64` got removed for now until the rust base port supports it.

**Comment by herbygillot** [comment]
> Edit: `supported_archs: arm64` got removed for now until the rust base port supports it.

Makes perfect sense. Merging.

**Pull request merged**